### PR TITLE
Assignment 7+8 linq - Zach Lawson and Tom Rohr

### PR DESCRIPTION
**Pull request by zlawson-ewu**
Assignment 7+8 LINQ peer programmed by Zach Lawson (@zlawson-ewu) and Tom Rohr (@Tom-Rohr). No extra credit attempted.

### CODE REVIEW

**Comment by jstanley6** [comment]
I will review this

**Review by jstanley6 — COMMENTED** [review]
1. Implement the ISampleData.CsvRows property, loading the data from the People.csv file and returning each line as a single string. ✔

- Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed 
  along with your test project. ✔
- Using LINQ, skip the first row in the People.csv. ✔
- Be sure to appropriately handle resource (IDisposable) items correctly if applicable (and it may 
  not be depending on how you implement it). ✔

2. Implement IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows() to return a sorted, unique list of states. ✔

- Use ISampleData.CsvRows for your data source. ✔
- Don't forget the list should be unique. ✔
- Sort the list alphabetically. ✔
- Include a test that leverages a hardcoded list of addresses. ✔
- Include a test that uses LINQ to verify the data is sorted correctly (do not use a hardcoded 
   list). ✔

3. Implement ISampleData.GetAggregateSortedListOfStatesUsingCsvRows() to return a string that contains a unique, comma separated list of states. ✔

- Use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows() for your data source. ✔
- Consider "selecting" only the states and calling ToArray() to retrieve an array of all the state names. ✔
- Given the array, consider using string.Join to combine the list into a single string. ✔

4. Implement the ISampleData.People property to return all the items in People.csv as Person objects ✔

- Use ISampleData.CsvRows as the source of the data. ✔
- Sort the list by State, City, and Zip. (Sort the addresses first then select). ❌
- Be sure that Person.Address is also populated. ✔
- Adding null validation to all the Person and Address properties is optional.
- Consider using ISampleData.CsvRows in your test to verify your results. ✔

5. Implement ISampleDate.FilterByEmailAddress(Predicate<string> filter) to return a list of names where the email address matches the filter. ✔

- Use ISampleData.People for your data source. ✔

6. Implement ISampleData.GetAggregateListOfStatesGivenPeopleCollection(IEnumerable<IPerson> people) to return a string that contains a unique, comma-separated list of states. ✔

- Use the people parameter from ISampleData.People property for your data source. ✔
- At a minimum, use the System.Linq.Enumerable.Aggregate` LINQ method to create your result. ✔
- Don't forget the list should be unique. ✔
- It is recommended that, at a minimum, you use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows to validate your result.

7. Given the implementation of Node in Assignment5

- Implement IEnumerable<T> to return all the items in the "circle" of items. ✔
- Add an IEnumberable<T> ChildItems(int maximum) method to Node that returns the remaining items with a maximum number of items returned less than maximum.

Fundamentals
Place all shared project properties into a Directory.Build.Props file. ✔
Place all shared project items into a Directory.Build.targets file.
Ensure nullable reference types is enabled ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
and enabled .NET analyzers for both projects ✔
For this assignment, consider using Assert.AreEqual<T>() (the generic version) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔

Good job the only thing that I found was 
Sort the list by State, City, and Zip. (Sort the addresses first then select).
I will add notes on how to fix this

File: Assignment/Assignment/SampleData.cs [hunk 1]
```diff
@@ -35,9 +36,19 @@ public string GetAggregateSortedListOfStatesUsingCsvRows()
...
+            return new Person(attributes[1].Trim(), attributes[2].Trim(), address, attributes[3].Trim());
+        }
         // 4.
-        public IEnumerable<IPerson> People
-            => CsvRows.Select(row => Person.ParseRow(row));
+        public IEnumerable<IPerson> People => CsvRows.Select(ParsePersonFromRow)
+            .OrderBy(x => x.Address.ToString())
+            .ThenBy(x => x.Address.State)
+            .ThenBy(x => x.Address.City)
+            .ThenBy(x => x.Address.Zip);
```
Comment (jstanley6) [inline Assignment/Assignment/SampleData.cs]: I would recommend using ".OrderBy().ThenBy().ThenBy().Select()" to sort the list before selecting each row to make the Person object

Comment (zlawson-ewu) [reply Assignment/Assignment/SampleData.cs]: Fixed that, thank you. I think I misinterpreted that bullet point.

File: Assignment/Assignment/SampleData.cs [hunk 2]
```diff
@@ -1,30 +1,70 @@
...
-        // 6.
-        public string GetAggregateListOfStatesGivenPeopleCollection(
-            IEnumerable<IPerson> people) => throw new NotImplementedException();
+    // 5.
+    public IEnumerable<(string FirstName, string LastName)> FilterByEmailAddress(
+        Predicate<string> filter) 
+    {
+        List<IPerson> matches = People.Where(x => filter(x.EmailAddress)).ToList();
+        List<(string, string)> nameList = new();
+        foreach (IPerson item in matches)
```
Comment (mmwoodfo) [inline Assignment/Assignment/SampleData.cs:51]: You should use select here!

File: Assignment/Assignment/SampleData.cs [hunk 3]
```diff
@@ -1,30 +1,70 @@
...
+        }
+        return nameList;
+    }
+
+    // 6.
+    public string GetAggregateListOfStatesGivenPeopleCollection(
+        IEnumerable<IPerson> people)
+    {
+        List<string> states = new();
+        foreach (IPerson item in people)
```
Comment (mmwoodfo) [inline Assignment/Assignment/SampleData.cs:63]: Again, avoid foreach, favor using linq methods

**Pull request closed without merging**